File: web/src/lib/agents/tools.ts
// Définitions des outils (skills) exposés à l'agent vocal, au format Vapi
// (JSON Schema). L'exécution réelle est dans src/lib/skills/*, appelée par
// /api/vapi/webhook. Un outil = une capacité ; un service externe = un module.
// Les descriptions sont en anglais : elles sont lues par le LLM (qui gère
// aussi bien les appels FR que EN), pas par l'utilisateur.

import { webhookServer } from "../vapi";

function serverTool(
  name: string,
  description: string,
  properties: Record<string, unknown>,
  required: string[] = [],
) {
  return {
    type: "function" as const,
    async: false,
    function: {
      name,
      description,
      parameters: { type: "object", properties, required },
    },
    server: webhookServer(),
  };
}

export function agentTools() {
  return [
    // --- Agenda (Google Calendar) ---
    serverTool(
      "list_events",
      "List the user's calendar events for a given day.",
      { day: { type: "string", description: "Requested day as YYYY-MM-DD, or 'today'/'tomorrow'" } },
      ["day"],
    ),
    serverTool(
      "create_event",
      "PROPOSE a new calendar event. With confirmed=false it returns a proposal to read out loud. Only pass confirmed=true after an explicit spoken 'yes' from the user.",
      {
        title: { type: "string", description: "Event title" },
        start: { type: "string", description: "Start in ISO 8601, e.g. 2026-07-15T14:00:00" },
        duration_minutes: { type: "number", description: "Duration in minutes (default 60)" },
        confirmed: { type: "boolean", description: "true only after explicit spoken confirmation" },
      },
      ["title", "start", "confirmed"],
    ),
    serverTool(
      "move_event",
      "PROPOSE to move an existing event. Same confirmation rule as create_event.",
      {
        event_query: { type: "string", description: "Words from the title or the time of the event to move" },
        new_start: { type: "string", description: "New start, ISO 8601" },
        confirmed: { type: "boolean" },
      },
      ["event_query", "new_start", "confirmed"],
    ),

    // --- Rappels ---
    serverTool(
      "set_reminder",
      "Schedule a reminder to be sent by SMS at the requested time.",
      {
        text: { type: "string", description: "Reminder text, e.g. 'take the bread out of the oven'" },
        due_at: { type: "string", description: "When, ISO 8601" },
        recurrence: { type: "string", enum: ["none", "daily", "weekly", "monthly"], description: "Recurrence" },
      },
      ["text", "due_at"],
    ),
    serverTool("list_reminders", "List the user's upcoming reminders.", {}),
    serverTool(
      "did_i_already",
      "Answer 'did I already do X today?' by checking reminders marked as done.",
      { what: { type: "string", description: "The action asked about, e.g. 'watered the plants'" } },
      ["what"],
    ),
    serverTool(
      "mark_done",
      "Mark an action/reminder as done (e.g. the user says they just did it).",
      { what: { type: "string" } },
      ["what"],
    ),

    // --- Météo ---
    serverTool(
      "get_weather",
      "Give the weather for a city (today or tomorrow).",
      {
        city: { type: "string", description: "City. If absent, use the user's home city." },
        day: { type: "string", enum: ["today", "tomorrow"], description: "Day" },
      },
    ),
    serverTool(
      "get_current_time",
      "Give the current local time in a city.",
      {
      city: {
      type: "string",
      description: "City whose current local time is requested",
    },
  },
  ["city"],
),

    // --- Navigation par SMS ---
    serverTool(
      "get_directions",
      "Compute a route and text the steps to the user by SMS, in addition to the spoken summary.",
      {
        destination: { type: "string", description: "Destination address or place" },
        origin: { type: "string", description: "Starting point. If absent: ask 'where are you?' or use the home address." },
        mode: { type: "string", enum: ["walking", "driving", "transit"], description: "Travel mode" },
      },
      ["destination"],
    ),

    // --- Contacts ---
    serverTool(
      "find_contact",
      "Search a contact (phone number) in the user's Google address book.",
      { name: { type: "string", description: "Contact name or nickname" } },
      ["name"],
    ),

    // --- Messages ---
    serverTool(
      "send_sms",
      "PROPOSE to send a dictated SMS. With confirmed=false: returns the text to read back out loud. confirmed=true only after an explicit 'yes'. Sensitive action: requires the verified PIN.",
      {
        to_name: { type: "string", description: "Recipient name (resolved via contacts)" },
        to_number: { type: "string", description: "Or a direct E.164 number" },
        body: { type: "string", description: "Dictated message text" },
        confirmed: { type: "boolean" },
      },
      ["body", "confirmed"],
    ),

    // --- Appels sortants (Rendez-vous / Taxi / Résa) ---
    serverTool(
      "place_call",
      "PROPOSE to place a call on the user's behalf (book an appointment, a taxi, a restaurant…). With confirmed=false: returns a mission recap to read out loud. confirmed=true only after an explicit 'yes'. Sensitive action: requires the verified PIN. The result will be sent by SMS.",
      {
        kind: {
          type: "string",
          enum: ["appointment", "taxi", "resto", "generic"],
          description:
            "Mission preset. 'appointment' = any appointment booking (doctor, hairdresser, garage…), 'taxi' = taxi booking, 'resto' = restaurant booking, 'generic' = anything else.",
        },
        goal: { type: "string", description: "Precise goal, e.g. 'book a haircut this week, mornings preferred'" },
        target_name: { type: "string", description: "Name of the place/person to call" },
        target_number: { type: "string", description: "Number to call if known (otherwise looked up in contacts/memory)" },
        constraints: { type: "string", description: "Useful constraints: time slots, number of people, pickup address…" },
        confirmed: { type: "boolean" },
      },
      ["kind", "goal", "confirmed"],
    ),

    // --- Mémoire ---
    serverTool(
      "remember",
      "Store a durable fact about the user (e.g. 'my mechanic is Garage Dupont, 01 23 45 67 89').",
      {
        key: { type: "string", description: "Short topic, e.g. 'mechanic'" },
        value: { type: "string", description: "The information to remember" },
      },
      ["key", "value"],
    ),
    serverTool(
      "recall",
      "Search the user's memory (places, people, preferences already noted).",
      { query: { type: "string", description: "What to look for" } },
      ["query"],
    ),

    // --- Sécurité ---
    serverTool(
      "verify_pin",
      "Verify the user's spoken PIN. Required before any sensitive action (sending a message, placing a call). NEVER repeat the PIN out loud.",
      { pin: { type: "string", description: "The dictated 4-digit code" } },
      ["pin"],
    ),
  ];
}

File: runtime/tools.py
"""Outils exposés au LLM — miroir exact de web/src/lib/agents/tools.ts.

Les schémas sont déclarés ici (format Pipecat), mais l'EXÉCUTION est
entièrement déléguée à l'API Next.js (/api/tools/execute) : une seule
implémentation des skills, quel que soit le runtime vocal.
"""

from collections.abc import Awaitable, Callable

from pipecat.adapters.schemas.function_schema import FunctionSchema
from pipecat.adapters.schemas.tools_schema import ToolsSchema
from pipecat.services.llm_service import FunctionCallParams

import api_client


def _schema(name: str, description: str, properties: dict, required: list[str]) -> FunctionSchema:
    return FunctionSchema(name=name, description=description, properties=properties, required=required)


END_CALL = _schema(
    "end_call",
    "Termine l'appel poliment (après avoir dit au revoir).",
    {},
    [],
)


def inbound_tools() -> ToolsSchema:
    """Outils de l'agent entrant (assistant personnel)."""
    return ToolsSchema(
        standard_tools=[
            _schema(
                "list_events",
                "Liste les rendez-vous de l'agenda de l'utilisateur pour un jour donné.",
                {
                    "day": {
                        "type": "string",
                        "description": "Jour au format AAAA-MM-JJ, ou 'today'/'tomorrow'",
                    }
                },
                ["day"],
            ),
            _schema(
                "create_event",
                "PROPOSE un nouveau rendez-vous. confirmed=false renvoie une proposition à lire "
                "à voix haute ; confirmed=true seulement après un 'oui' explicite.",
                {
                    "title": {"type": "string"},
                    "start": {"type": "string", "description": "Début ISO 8601"},
                    "duration_minutes": {"type": "number"},
                    "confirmed": {"type": "boolean"},
                },
                ["title", "start", "confirmed"],
            ),
            _schema(
                "move_event",
                "PROPOSE de déplacer un rendez-vous existant. Même règle de confirmation.",
                {
                    "event_query": {"type": "string"},
                    "new_start": {"type": "string"},
                    "confirmed": {"type": "boolean"},
                },
                ["event_query", "new_start", "confirmed"],
            ),
            _schema(
                "set_reminder",
                "Programme un rappel envoyé par SMS au moment voulu.",
                {
                    "text": {"type": "string"},
                    "due_at": {"type": "string", "description": "ISO 8601"},
                    "recurrence": {"type": "string", "enum": ["none", "daily", "weekly", "monthly"]},
                },
                ["text", "due_at"],
            ),
            _schema("list_reminders", "Liste les rappels à venir.", {}, []),
            _schema(
                "did_i_already",
                "Répond à « est-ce que j'ai déjà fait X ? » via les rappels marqués faits.",
                {"what": {"type": "string"}},
                ["what"],
            ),
            _schema("mark_done", "Marque une action comme faite.", {"what": {"type": "string"}}, ["what"]),
            _schema(
                "get_weather",
                "Météo d'une ville (aujourd'hui ou demain).",
                {"city": {"type": "string"}, "day": {"type": "string", "enum": ["today", "tomorrow"]}},
                [],
            ),
            _schema(
                "get_directions",
                "Itinéraire : résumé vocal + étapes envoyées par SMS.",
                {
                    "destination": {"type": "string"},
                    "origin": {"type": "string"},
                    "mode": {"type": "string", "enum": ["walking", "driving", "transit"]},
                },
                ["destination"],
            ),
            _schema(
                "find_contact",
                "Cherche un contact dans le carnet Google.",
                {"name": {"type": "string"}},
                ["name"],
            ),
            _schema(
                "send_sms",
                "PROPOSE un SMS dicté (relecture puis confirmed=true). Protégé : code requis "
                "(request_code puis verify_code).",
                {
                    "to_name": {"type": "string"},
                    "to_number": {"type": "string"},
                    "body": {"type": "string"},
                    "confirmed": {"type": "boolean"},
                },
                ["body", "confirmed"],
            ),
            _schema(
                "place_call",
                "PROPOSE un appel passé à la place de l'utilisateur (rendez-vous, taxi, resto). "
                "Récapitulatif puis confirmed=true. Protégé : code requis (request_code puis "
                "verify_code). Résultat par SMS.",
                {
                    "kind": {"type": "string", "enum": ["appointment", "taxi", "resto", "generic"]},
                    "goal": {"type": "string"},
                    "target_name": {"type": "string"},
                    "target_number": {"type": "string"},
                    "constraints": {"type": "string"},
                    "confirmed": {"type": "boolean"},
                },
                ["kind", "goal", "confirmed"],
            ),
            _schema(
                "remember",
                "Retient une information durable (ex: médecin traitant).",
                {"key": {"type": "string"}, "value": {"type": "string"}},
                ["key", "value"],
            ),
            _schema(
                "recall",
                "Recherche dans la mémoire de l'utilisateur.",
                {"query": {"type": "string"}},
                ["query"],
            ),
            _schema(
                "request_code",
                "Envoie un code à 4 chiffres par SMS au numéro enregistré. À appeler la première fois "
                "que l'utilisateur demande quelque chose de protégé (lire agenda/contacts/rappels/notes, "
                "modifier l'agenda, envoyer un SMS, passer un appel).",
                {},
                [],
            ),
            _schema(
                "verify_code",
                "Vérifie le code que l'utilisateur dit ou tape au clavier. Débloque les actions protégées "
                "pour le reste de l'appel. Ne jamais répéter le code à voix haute.",
                {"code": {"type": "string"}},
                ["code"],
            ),
            _schema(
                "get_current_time",
                "Donne l'heure locale actuelle dans une ville.",
                {
                    "city": {
                        "type": "string",
                        "description": "Ville dont l'heure locale est demandée",
                    }
                },
                ["city"],
            ),
            _schema(
                "define",
                "Donne la définition d'un mot anglais.",
                {"word": {"type": "string", "description": "Le mot anglais à définir"}},
                ["word"],
            ),
            _schema(
                "convert",
                "Convertit une valeur entre unités (distance, poids, température, volume, vitesse) "
                "ou entre devises (codes ISO comme EUR, USD).",
                {
                    "value": {"type": "number", "description": "La valeur numérique à convertir"},
                    "from": {
                        "type": "string",
                        "description": "Unité ou devise de départ, ex. 'km', 'kg', 'celsius', 'EUR'",
                    },
                    "to": {
                        "type": "string",
                        "description": "Unité ou devise d'arrivée, ex. 'miles', 'lb', 'fahrenheit', 'USD'",
                    },
                },
                ["value", "from", "to"],
            ),
            _schema(
                "report_unsupported_request",
                "À appeler UNE fois quand l'appelant demande quelque chose qu'AUCUN autre outil ne "
                "sait faire — une capacité manquante. Décris la CAPABILITÉ manquante en anglais, en "
                "termes généraux, pas les détails privés de l'appelant. notify_caller=true seulement "
                "si tu as proposé un SMS et que l'appelant a dit oui.",
                {
                    "request_summary": {"type": "string"},
                    "caller_words": {"type": "string"},
                    "language": {"type": "string", "enum": ["en", "fr"]},
                    "notify_caller": {"type": "boolean"},
                },
                ["request_summary"],
            ),
            END_CALL,
        ]
    )


def outbound_tools() -> ToolsSchema:
    """Outils des missions sortantes (rendez-vous / taxi / resto)."""
    return ToolsSchema(
        standard_tools=[
            _schema(
                "report_outcome",
                "Rapporte le résultat final de la mission (à appeler UNE fois avant de raccrocher). "
                "details sera envoyé tel quel par SMS au client : rédige-le pour lui.",
                {
                    "status": {"type": "string", "enum": ["success", "failed", "voicemail", "needs_user"]},
                    "details": {"type": "string"},
                },
                ["status", "details"],
            ),
            END_CALL,
        ]
    )


def make_tool_handler(
    call_id: str,
    job_id: str | None,
    hangup: Callable[[], Awaitable[None]],
) -> Callable[[FunctionCallParams], Awaitable[None]]:
    """Handler attrape-tout : forwarde chaque appel d'outil vers l'API Next."""

    async def handler(params: FunctionCallParams) -> None:
        if params.function_name == "end_call":
            await params.result_callback("Appel terminé.")
            await hangup()
            return
        try:
            result = await api_client.execute_tool(
                call_id, params.function_name, params.arguments or {}, job_id
            )
        except Exception as err:  # noqa: BLE001 — l'agent doit rester en ligne
            result = f"Désolé, ce service ne répond pas ({type(err).__name__})."
        await params.result_callback(result)

    return handler
